bip-package-relay.mediawiki

```diff
@@ -0,0 +1,241 @@
+
+===New Messages===
+
+Three new protocol messages are added for use in any version of package relay. Additionally, each
+version of package relay must define its own inv type and "pckginfo" message version, referred to in
+this document as "MSG_PCKG" and "pckginfo" respectively. See BIP-v1-packages for a concrete example.
+
+====sendpackages====
+
+{|
+|  Field Name  ||  Type  ||  Size  ||  Purpose
+|-
+|version || uint32_t || 4 || Denotes a package version supported by the node.
+|-
+|max_count || uint32_t || 4 ||Specifies the maximum number of transactions per package this node is willing to accept.
+|-
+|max_weight || uint32_t || 4 ||Specifies the maximum total weight per package this node is willing to accept.
+|-
+|}
+
+# The "sendpackages" message has the structure defined above, with pchCommand == "sendpackages".
+
+# During version handshake, nodes should send a "sendpackages" message indicate they support package relay and may request packages.
+
+# The message should contain a version supported by the node. Nodes should send a "sendpackages" message for each version they support.
+
+# The "sendpackages" message MUST be sent before sending a "verack" message. If a "sendpackages" message is received afer "verack", the sender should be disconnected.
+
+# If 'fRelay==false' in a peer's version message, the node must not send "sendpackages" to them. If a "sendpackages" message is received by a peer after sending `fRelay==false` in their version message, the sender should be disconnected.
+
+# Upon receipt of a "sendpackages" message with a version that is not supported, a node must treat the peer as if it never received the message.
+
+# If both peers send "wtxidrelay" and "sendpackages" with the same version, the peers should announce, request, and send package information to each other.
+
+====getpckgtxns====
+
+{|
+|  Field Name  ||  Type  ||  Size  ||   Purpose
+|-
+|txns_length||CompactSize||1 or 3 bytes|| The number of transactions requested.
+|-
+|txns||List of wtxids||txns_length * 32|| The wtxids of each transaction in the package.
+|}
+
+# The "getpckgtxns" message has the structure defined above, with pchCommand == "getpckgtxns".
+
+# A "getpckgtxns" message should be used to request all or some of the transactions previously announced in a "pckginfo" message, specified by witness transactiosome id.
+
+# Upon receipt of a "getpckgtxns" message, a node must respond with either a "pckgtxns" containing the requested transactions or a "notfound" message indicating one or more of the transactions is unavailable. This allows the receiver to avoid downloading and storing transactions that cannot be validated immediately.
+
+# A "getpckgtxns" message should only be sent if both peers agreed to send packages in the version handshake. If a "getpckgtxns" message is received from a peer with which package relay was not negotiated, the sender should be disconnected.
```

Comment:
Trying to think if disconnect is the right response here rather than merely ignore and not disconnect. Sending it (when no negotiation has taken place) does seem to suggest at the very least that the P2P logic of the node is faulty which I guess is reason to disconnect.